### DISCUSSION TITLE
Difference in function boundary condition solution to expected values

### DISCUSSION THREAD

**Discussion by AmbroiseJuston** [idaholab/moose#28735] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Good afternoon,

I am working on a problem which required me changing a boundary condition to simulate different test conditions. Basically i am doing an IV plot of an electrical component. 

We were trying to validate which way would better represent sweeping the IV plane (from 0v to 1v) and had some questions on why certain results diverged. 

Option 1: we can manually solve for steady state (~10 units of time) and then use the final state as the IC of the next simulation, increasing the voltage by a certain amount (0.1 for example).
Option 2: we can used a piecewise function and create a discrete stair function with increments of 0.1 after x time (~10 units of time to let it each steady state at that condition.
Option 3: create another piecewise function but linearly increasing from 0 to 1 and then back down.

I can see how option 3 would be different that then other solutions however what i do not understand if how options 1 and 2 diverge because in my mind they are equivalent in the long run. The difference is if we do the step manually or as a discrete step function.

Is there a difference between how MOOSE would evaluate the results for options 1 and 2?

The step function being used is shown below:
```
   [./bc_function_step]
    type = PiecewiseLinear
    y = '0 0 0.001  0.001  0.01  0.01  0.1   0.1  0.2   0.2  0.3   0.3  0.4  0.4  0.5   0.5  0.6  0.6  0.7   0.7 0.8  0.8 0.9   0.9 1      1'
    x = '0 5 5.001  15    15.001  25  25.001  35 35.001 45  45.001 55  55.001 65 65.001 75  75.001 85 85.001 95 95.001 105 105.001 115 115.001 125'
    scale_factor = 1.0
  [../]
```
and the code that reads in the .e as the new IC for the simulation is defined here:
```
[UserObjects]
  [exo_soln]
    type = SolutionUserObject
    mesh = 'etpf-23_test_out_IC23.e'
    system_variables = 'c'
    timestep = LATEST
  []
[]
```

**Comment by GiudGiud** [comment]
Hello

Are you using pseudo timesteps in all these simulations? So timesteps but with no time derivative kernel so a steady state is reached on every time step?

Guillaume

**Reply by AmbroiseJuston** [reply]
there is a time derivative kernel included and we are running until it reaches to steady-state if that was the question

**Reply by GiudGiud** [reply]
I could see the time integration error be different in option 1 and 2. 
how different are they? And how do you know the solution is at steady state after 10 units of time? did you use the built-in steady state detection?

**Reply by AmbroiseJuston** [reply]
They are off by around 80-120%, depending on the voltage at the step.

We have data showing that once it runs for around 5-10 ns of simulation time, the current (the quantity of interest) in the system reaches an equilibrium. To make sure, we are running for 20-30ns of simulation time. We know there is a built-in steady state solver, but this is mostly reusing the code of when we were interested in running for a certain amount of time. We also wanted to have more control over when the application ends while doing the sweep.

**Reply by GiudGiud** [reply]
how are you comparing the two solutions? visually using the exodus file at the last time step (that part is important of course)

>  We know there is a built-in steady state solver

it should be as easy as setting the executioner type to "Steady" and passing "--allow-unused" on the command line for the unused executiner parameters.
This might tell you which solution has unexpected results

**Reply by AmbroiseJuston** [reply]
Both visually and we are comparing the current that passes through the sample. The conductivity are comparable between options. The biggest difference is that difference in steady state current which should be same same if the conditions are identical between the options.

We can't just move it to a steady solve because one of our kernels on chemical potential uses some features from the transient solver.

**Reply by GiudGiud** [reply]
@cticenhour do you see why that could be? Or can you think of someone who does electronics w/ moose?

**Reply by AmbroiseJuston** [reply]
For some clarification, we are using phase-field for this type of analysis in case that helps

**Reply by GiudGiud** [reply]
so phase field + electronics.  Is phase field to compute the conductivities / material properties?

I wonder if @mangerij could help us here?
@csdechant do you know about this field?

**Reply by AmbroiseJuston** [reply]
Phase field is used for the analysis of memristors. For some more context, we also ran some tests with the energy function kernel to see if the problem was it was not actually reaching a system steady state, but it seems to. Also, when stepping the BC in voltage, there is no accompanying increase in the energy function at the same time as (we think) we should expect the same is true when doing it by hand it seems.

**Reply by GiudGiud** [reply]
so you're saying you are finding a missing coupling effect?
How is the coupling performed? In a single-system approach or with multiapps?
If in a single system approach, any auxiliary variable involved in the coupling ? (these can introduce lag)
If a multiapp approach, did you turn on fixed point iterations?

**Reply by AmbroiseJuston** [reply]
apologies for the delay.

The coupling is done in a single system. We are looking into multiapps for parameter studies but that is not what is being done here. There are a few auxiliary variable. Can these be retreived into the user object? If so i am not sure how.

**Reply by GiudGiud** [reply]
> There are a few auxiliary variables

are they used now in the coupling? If so, how are the auxkernels set up? Are they executed often enough to keep the variables updated?


>. Can these be retreived into the user object? If so i am not sure how.

Yes. Depending on the user object type there are different ways but for ElementUserObjects for example (a base class of many user objects), you can use the Coupleable interface

https://mooseframework.inl.gov/source/interfaces/Coupleable.html

**Reply by AmbroiseJuston** [reply]
I will go and double check how often they are being updated. From what i remember it should be often enough but i will make sure. We are using "SolutionUserObject" is there any difference in syntax between using apostrophes or just using spaces as shown below:

[UserObjects]
  [exo_soln]
    type = SolutionUserObject
    mesh = 'etpf-23_test_out_IC23.e'
    system_variables = 'var1 var2 var3 aux_var1 aux_var2 aux_var3'
    timestep = LATEST
  []
[]

Versus

[UserObjects]
  [exo_soln]
    type = SolutionUserObject
    mesh = 'etpf-23_test_out_IC23.e'
    system_variables = var1 var2 var3 aux_var1 aux_var2 aux_var3
    timestep = LATEST
  []
[]

**Reply by GiudGiud** [reply]
I would think it errors if you do : 

param = var1 var2 var3

But if it's being accepted then no no difference

**Reply by AmbroiseJuston** [reply]
I have been playing around, and it seems like the error might be caused by some variables not being properly ported over (because of user error). I will play around with it for some more to see if there are any more major mistakes I am making.

you will probably hear from me again but in the mean time I apologise for my slow communication

**Reply by GiudGiud** [reply]
no worries. By error you mean the difference in behavior still right, not an error from the code.
